Source/WebKit/Shared/RTCPacketOptions.cpp

```diff
@@ -34,66 +34,35 @@
+    options.packet_id = data.packetId;
 
-    encoder << static_cast<int64_t>(options.packet_time_params.srtp_auth_tag_len);
-    if (options.packet_time_params.srtp_auth_tag_len > 0)
-        encoder << IPC::DataReference(reinterpret_cast<const uint8_t*>(options.packet_time_params.srtp_auth_key.data()), options.packet_time_params.srtp_auth_key.size());
-
-    encoder << options.packet_time_params.srtp_packet_index;
-}
-
-std::optional<RTCPacketOptions> RTCPacketOptions::decode(IPC::Decoder& decoder)
-{
     rtc::PacketTimeUpdateParams params;
-    rtc::PacketOptions options;
-
-    rtc::DiffServCodePoint dscp;
-    if (!decoder.decode(dscp))
-        return std::nullopt;
-    options.dscp = dscp;
-
-    std::optional<int32_t> packetId;
-    decoder >> packetId;
-    if (!packetId)
-        return std::nullopt;
-    options.packet_id = packetId.value();
-
-    std::optional<int> rtpSendtimeExtensionId;
-    decoder >> rtpSendtimeExtensionId;
-    if (!rtpSendtimeExtensionId)
-        return std::nullopt;
-    params.rtp_sendtime_extension_id = rtpSendtimeExtensionId.value();
-
-    std::optional<int64_t> srtpAuthTagLength;
-    decoder >> srtpAuthTagLength;
-    if (!srtpAuthTagLength)
-        return std::nullopt;
-    params.srtp_auth_tag_len = srtpAuthTagLength.value();
-
-    if (params.srtp_auth_tag_len > 0) {
-        IPC::DataReference srtpAuthKey;
-        if (!decoder.decode(srtpAuthKey))
-            return std::nullopt;
-
-        params.srtp_auth_key = std::vector<char>(static_cast<size_t>(srtpAuthKey.size()));
-        memcpy(params.srtp_auth_key.data(), reinterpret_cast<const char*>(srtpAuthKey.data()), srtpAuthKey.size() * sizeof(char));
+    params.rtp_sendtime_extension_id = data.rtpSendtimeExtensionId;
+    params.srtp_auth_tag_len = data.srtpAuthTagLength;
+    if (data.srtpAuthTagLength > 0) {
+        auto* begin = reinterpret_cast<const char*>(data.srtpAuthKey.data());
+        params.srtp_auth_key = std::vector<char>(begin, begin + static_cast<size_t>(data.srtpAuthKey.size()));
     }
-
-    std::optional<int64_t> srtpPacketIndex;
-    decoder >> srtpPacketIndex;
-    if (!srtpPacketIndex)
-        return std::nullopt;
-    params.srtp_packet_index = srtpPacketIndex.value();
+    params.srtp_packet_index = data.srtpPacketIndex;
 
     options.packet_time_params = WTFMove(params);
-    return RTCPacketOptions { WTFMove(options) };
 }
 
+auto RTCPacketOptions::serializableData() const -> SerializableData
+{
+    return {
+        static_cast<int32_t>(options.dscp),
+        safeCast<int32_t>(options.packet_id),
+        options.packet_time_params.rtp_sendtime_extension_id,
+        static_cast<int64_t>(options.packet_time_params.srtp_auth_tag_len),
+        options.packet_time_params.srtp_auth_tag_len > 0  ? std::span<const uint8_t> { } : std::span<const uint8_t> { reinterpret_cast<const uint8_t*>(options.packet_time_params.srtp_auth_key.data()), options.packet_time_params.srtp_auth_key.size() },
```

Comment:
It is suspicious that srtp_auth_tag_len is serialized separately, then the length of this is not directly related to that value.  Do we need a bounds check when decoding srtp_auth_tag_len?

Reply:
I'm trying to match the behavior of the previous encoder / decoder.

Reply:
I don't think there is a security concern here. While we use srtp_auth_tag_len to decide whether or not to decode the span, the span has its own side and its decoding doesn't rely on srtp_auth_tag_len. I think matching the old coders like I did is OK here?